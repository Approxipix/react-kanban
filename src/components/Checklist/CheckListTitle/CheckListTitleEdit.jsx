import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { editCheckListTitle } from '../../../redux/checkListReducer/actions';
import ClickOutside from "../../ClickOutside";
import { Input, } from '../../BaseComponent';

class CheckListTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.checkListTitle || '',
    };
  }

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.props.toggleEditing()
    }
  };

  handleChange = (value) => {
    this.setState({
      title: value,
    });
  };


  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    const { actions } = this.props;
    if (!title)  return;
    actions.editCheckListTitle({
      checkListId: this.props.checkListId,
      checkListTitle: title,
    });
    this.props.toggleEditing()
  };

  render = () => {
    const { title } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditing}>
        <form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            padding=".3rem .5rem"
            margin="0"
            size="1.2rem"
            weight="600"
            value={title}
            placeholder="Enter list title..."
            onKeyDown={this.handleKeyDown}
            onChange={(e) => this.handleChange(e.target.value)}
            onBlur={this.handleSubmit}
            spellCheck={false}
            autoFocus
          />
        </form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editCheckListTitle: editCheckListTitle,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CheckListTitleEdit);
