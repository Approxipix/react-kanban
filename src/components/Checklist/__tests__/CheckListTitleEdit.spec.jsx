import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CheckListTitleEdit from '../CheckListTitle/CheckListTitleEdit';
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

describe('<CheckListTitleEdit>', () => {
  let store;
  let props;
  let component;
  let mockEditCheckListTitle = jest.fn();
  let mockToggleEditTitle = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();

    store = mockStore({});
    store.dispatch = mockEditCheckListTitle;
    props = {
      checkListId: 'CheckListID',
      checkListTitle: 'CheckList Title',
      toggleEditTitle: mockToggleEditTitle
    };

    const wrapper = shallow(<CheckListTitleEdit store={store} {...props} />);
    component = wrapper.find('CheckListTitleEdit').dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should call toggleEditTitle if key code equal 27", () => {
    component.find('Input').simulate('keydown', { keyCode: 27 });

    expect(mockToggleEditTitle).toHaveBeenCalledTimes(1);
  });

  it('should call handleChange on title change with the correct params', () => {
    const spy = jest.spyOn(component.instance(), "handleChange");
    const mockEvent = {
      target: {
        name: "title",
        value: "Test title"
      }
    };

    component.find("Input").simulate("change", mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(mockEvent);
  });

  it('should call setState on title', () => {
    const mockEvent = {
      target: {
        name: "title",
        value: "Test Title"
      }
    };
    const expectedState = {
      title: "Test Title",
    };

    component.instance().handleChange(mockEvent);

    expect(component.state()).toEqual(expectedState)
  });

  it("should call handleSubmit with the correct params", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.setState({ title: "Test Title" });
    component.instance().handleSubmit(mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should return handleSubmit if title is empty", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.instance().handleSubmit(mockEvent);

    expect(spy).toReturn();
  });

  it("should dispatch an action after handleSubmit", () => {
    const expectedAction = {
      type: 'EDIT_CHECKLIST_TITLE',
      payload: {
        checkListId: 'CheckListID',
        checkListTitle: 'Test Title',
      }
    };

    component.setState({ title: "Test Title" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockEditCheckListTitle).toHaveBeenCalledTimes(1);
    expect(mockEditCheckListTitle).toHaveBeenCalledWith(expectedAction);
  });

  it("should call toggleEditTitle after submit", () => {
    component.setState({ title: "Test Title" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockToggleEditTitle).toHaveBeenCalledTimes(1);
  });
});
