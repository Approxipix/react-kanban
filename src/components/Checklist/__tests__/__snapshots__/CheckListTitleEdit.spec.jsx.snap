// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckListTitleEdit> should render with given state from Redux store 1`] = `
<ClickOutside
  toggleOpened={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <Input
      autoFocus={true}
      margin="0"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      padding=".3rem .5rem"
      placeholder="Enter list title..."
      size="1.2rem"
      spellCheck={false}
      type="text"
      value="CheckList Title"
      weight="600"
    />
  </form>
</ClickOutside>
`;
